### PR TITLE
new_audit: ensure proper origin isolation with COOP

### PR DESCRIPTION
**Pull request by sebastian9er**
**Summary**
Adding a new audit to Ligththouse, which detects deviations from an optimal COOP header deployment for origin isolation.

Part of a larger change to introduce more similar header deployments. @adamraine FYI

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -68,3 +68,6 @@ yarn.lock
 !**/types/**/*.d.ts
 
 page-functions-test-case*out*.js
+package.json
+yarn.lock
+cli/test/smokehouse/frontends/smokehouse-bin.js
```
Comment (adamraine) [inline .gitignore:73]: Can you revert the changes to this file, smokehouse-bin.js, package.json and yarn.lock?

**Review by adamraine — CHANGES_REQUESTED** [review]

File: core/audits/origin-isolation.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+          f => this.findingToTableItem(
+              f.directive, f.description,
+              str_(i18n.UIStrings.itemSeverityHigh))),
+      ...syntax.map(
+          f => this.findingToTableItem(
+              f.directive, f.description,
+              str_(i18n.UIStrings.itemSeverityLow))),
+    ];
+
+    return {score: violations.length || syntax.length > 1 ? 0 : 1, results};
```
Comment (adamraine) [inline core/audits/origin-isolation.js:123]:
```suggestion
    return {score: violations.length || syntax.length ? 0 : 1, results};
```

File: core/audits/origin-isolation.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import {Audit} from './audit.js';
+import {MainResource} from '../computed/main-resource.js';
+import * as i18n from '../lib/i18n/i18n.js';
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. "COOP" stands for "Cross-Origin-Opener-Policy". */
+  title: 'Ensure the proper usage of the COOP header to isolate the origin.',
+  /** Description of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. This is displayed after a user expands the section to see more. No character length limits. The last sentence starting with 'Learn' becomes link text to additional documentation. "COOP" stands for "Cross-Origin-Opener-Policy". */
+  description: 'Deployment of the COOP header allows isolation of the top-level document to not share a browsing context group with cross-origin documents. ' +
```
Comment (benschwarz) [inline core/audits/origin-isolation.js:15]: Both title and description include "COOP header", but neither say "Cross-Origin-Opener-Policy". I'd never heard of this header prior to reading this PR, sooo… maybe a good idea to revise the description.

Comment (sebastian9er) [reply core/audits/origin-isolation.js:15]: Ah, yeah, good point. I will fix it in the code as well, if @adamraine doesn't think it's too long and we should re-phrase it generally.

File: cli/test/smokehouse/frontends/smokehouse-bin.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env node
 /**
  * @license
- * Copyright 2019 Google LLC
+ * Copyright 2024 Google LLC
```
Comment (adamraine) [inline cli/test/smokehouse/frontends/smokehouse-bin.js:4]:
```suggestion
 * Copyright 2019 Google LLC
```

File: core/audits/origin-isolation.js (hunk 3 above)
Comment (adamraine) [reply core/audits/origin-isolation.js:15]: In the long form description, we should use "Cross-Origin-Opener-Policy (COOP)" to identify the acronym. Everywhere else it should be fine to use "COOP"

File: core/audits/origin-isolation.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {Audit} from './audit.js';
+import {MainResource} from '../computed/main-resource.js';
+import * as i18n from '../lib/i18n/i18n.js';
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. "COOP" stands for "Cross-Origin-Opener-Policy". */
+  title: 'Ensure the proper usage of the COOP header to isolate the origin.',
```
Comment (adamraine) [inline core/audits/origin-isolation.js:13]:
```suggestion
  title: 'Ensure proper origin isolation with COOP',
```

File: core/audits/origin-isolation.js (hunk 3 above)
Comment (sebastian9er) [reply core/audits/origin-isolation.js]: SGTM, thanks Adam!

File: core/audits/origin-isolation.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /**
+   * @param {string[]} coopHeaders
+   * @return {{score: number, results: LH.Audit.Details.TableItem[]}}
+   */
+  static constructResults(coopHeaders) {
+    const rawCoop = [...coopHeaders];
+    const allowedDirectives =
+        ['unsafe-none', 'same-origin-allow-popups', 'same-origin'];
```
Comment (adamraine) [inline core/audits/origin-isolation.js:89]: What about `noopener-allow-popups`?

Comment (sebastian9er) [reply core/audits/origin-isolation.js]: Added!

File: core/audits/origin-isolation.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {Audit} from './audit.js';
+import {MainResource} from '../computed/main-resource.js';
+import * as i18n from '../lib/i18n/i18n.js';
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. "COOP" stands for "Cross-Origin-Opener-Policy". */
```
Comment (adamraine) [inline core/audits/origin-isolation.js:12]:
```suggestion
  /** Title of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. "COOP" stands for "Cross-Origin-Opener-Policy" and should not be translated. */
```

File: core/audits/origin-isolation.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+  title: 'Ensure proper origin isolation with COOP',
+  /** Description of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. This is displayed after a user expands the section to see more. No character length limits. The last sentence starting with 'Learn' becomes link text to additional documentation. "COOP" stands for "Cross-Origin-Opener-Policy". */
```
Comment (adamraine) [inline core/audits/origin-isolation.js:14]:
```suggestion
  /** Description of a Lighthouse audit that evaluates the security of a page's COOP header for origin isolation. This is displayed after a user expands the section to see more. No character length limits. The last sentence starting with 'Learn' becomes link text to additional documentation. "COOP" stands for "Cross-Origin-Opener-Policy", neither should be translated. */
```

File: core/audits/origin-isolation.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  description: 'The Cross-Origin-Opener-Policy (COOP) can be used to isolate the top-level window from other documents such as pop-ups. [Learn more about deploying the COOP header.](https://web.dev/articles/why-coop-coep#coop)',
+  /** Summary text for the results of a Lighthouse audit that evaluates the COOP header for origin isolation. This is displayed if no COOP header is deployed. "COOP" stands for "Cross-Origin-Opener-Policy". */
```
Comment (adamraine) [inline core/audits/origin-isolation.js:16]:
```suggestion
  /** Summary text for the results of a Lighthouse audit that evaluates the COOP header for origin isolation. This is displayed if no COOP header is deployed. "COOP" stands for "Cross-Origin-Opener-Policy" and should not be translated. */
```

**Review by adamraine — APPROVED** [review]
